### ISSUE TITLE
Check that all the promises are caught correctly in cursor.mocha

### ISSUE THREAD

**Issue by nighelles** [rethinkdb/rethinkdb#5413]

**Comment by larkost** [comment]
Just clarifying this: while @nighelles and I were working on a new test he was creating within `cursor.mocha` I got worried that other existing tests were not `return`ing the promises that they were using, and so failures within those test could well go unmarked. This bug is to audit that to make sure we are doing the right thing.

**Comment by larkost** [comment]
I have an initial version of this pushed to branch `larkost/5413-javascript-connection-tests`. That winds up doing a lot more than this initially started out as it adds extensive testing for the new behavior for `eachAsync`, as well as as preliminary testing for #5478 (has not been checked with that code yet).

This does wind up turning up a trio of repeated problems that @nighelles is working on.

**Comment by marshall007** [comment]
@larkost I don't think [`cursor.mocha#L542-L560`](https://github.com/rethinkdb/rethinkdb/blob/331da0f8b37fcabbde9b86f8cbd249250fbf823e/test/rql_test/connections/cursor.mocha#L542-L560) is a valid test. Enforcing this behavior would require you to `return Promise.resolve()` as a no-op when you want to skip over a row, for example.

**Comment by nighelles** [comment]
Returning undefined is not an error, right?

On Tue, Mar 15, 2016 at 4:13 PM, Marshall Cottrell <notifications@github.com

**Comment by larkost** [comment]
Ya, just `return;` would be valid, just not returning anything other than a `Promise`. The goal is to make sure that people get an error message rather than us silently swallowing values.

**Comment by marshall007** [comment]
Ah, ok. As long as `return;` is still valid that seems reasonable.

**Comment by marshall007** [comment]
@larkost all the failures with `row error passed through outer promise rather than final function` are not valid. `eachAsync` will always return a promise and `.catch(...)` is now the preferred way to handle exceptions. The (optional) finish callback is only there for callback-style people who aren't looking at the promise returned from `eachAsync` and, if specified, should be called **in addition to** the promise being rejected.

**Comment by marshall007** [comment]
@larkost also in [`cursor.mocha#L929-L956`](https://github.com/rethinkdb/rethinkdb/blob/331da0f8b37fcabbde9b86f8cbd249250fbf823e/test/rql_test/connections/cursor.mocha#L929-L956). This is not intended behavior. As I mentioned, the final callback is only for the callback-style and should not be used in conjunction with promises. Expecting a promise returned from the final callback to be resolved before the `.then(...)` doesn't make sense because you shouldn't be using both.

```js
// valid callback-style
cursor.eachAsync(function (row, cb) {
  return cb('error');
}, function (err) {
  // err == 'error'
})

// valid promise-style
cursor.eachAsync(function (row) {
  return Promise.reject('error');
})
.then(function () {
  // ...
})
.catch(function (err) {
  // err == 'error'
})
```

The tests should be written in a way that that reflects one of these styles, not both, as many tests currently do:

```js
// mixing both styles
cursor.eachAsync(function (row) {
  return Promise.reject('error');
}, function (err) {
  // err == 'error'
})
.then(function () {
  // ...
})
```

**Comment by larkost** [comment]
@marshall007: I disagree with you on both of your previous posts:
1. That `eachAsync` returns a promise even when there is a `final` function (callback style) already demonstrates that we are going to support "mixed" behavior. Since we have gone down that road we need to embrace it, not ignore things because they are not pure.
2. I am strongly against sending an error both to the `final` handler and out the promise direction. In the case of people who are trying to use the callback style this would leave them with unhanded promise rejections if they did not also put in a `.catch` block, and render the whole `final` block necessarily redundant. So when people give a `final` handler, it should be the only one to deal with errors encounter during `row` handler processing. I see room for debate about whether errors thrown in the `final` handler should be thrown generally, or sent out through a Promise rejection.
3. If we are going to wait for resolution on promises from the `row` hander, then we should also wait on promises from the `final` hander. Doing otherwise seems to me to be sacrificing both convince and predictability for the sake of purity. If we were not to accept/resolve a promise there, then we need to make it an error to prevent subtle bugs from creeping into people's code.
4. And the tests should most definitely test the mixed cases (in addition to the straight-line ones). No matter what we decide the behavior should be, we need to maintain that behavior even in circumstances we don't intend for people to use. This is how we can guard against breaking people's code accidentally. Note that I am not advocating that we encourage people to use these mixed cases (I don't mind personally), but we need to be careful about maintaining our behavior.

**Comment by marshall007** [comment]
@larkost I agree with you on points (1) and (4). My main point is that they are two separate APIs and thus using one should not have side-effects on the other. As a result, I do not think handling the error in the `final` callback and/or returning a promise from there should have any impact on the outer promise.

**Comment by larkost** [comment]
@marshall007: But they are not separate APIs. There is only one call, and we branch our behavior on how you make it (first on number of functions, then again on arguments to the first function). We have to have some sort of logical construct that its over all uses. Can you clarify how your argument does not make the `final` function's handling of errors necessarily redundant (since you would always have to `.catch` the outer block anyways) for (2), and then why we should not treat Promises returned from a `final` handler just like we do one from the `row` handler for (3)?

**Comment by marshall007** [comment]
@larkost you do not necessarily have to catch the outer promise, it can resolve without error. If we were to have implemented `eachAsync` without callback support, the only supported syntax would be:

```js
cursor.eachAsync(function (row) {
  return Promise.reject('error');
})
.then(function () {
  // ...
})
.catch(function (err) {
  // err == 'error'
})
```

Since we also want to support the traditional node callback style, we are essentially "nodeifying" this API. So each part of the promise-based API has an equivalent in the callback-style, but the functionality does not interact.

```js
cursor.eachAsync(function (row, cb) {
  cb('error')      // return Promise.reject('error'); 
}, function (err) {
  if (err == null) // .then(function () { })
  else             // .catch(function (err) { })
})
```

Notice how returning a promise from `final` here wouldn't make sense. It would be like expecting something like this to work:

```js
r.table('foo').run(conn, (err, data) {
  return Promise.delay(10);
})
.then(function () {
  // delay will not have been resolved, which
  // is correct behavior since we used callbacks
})
```

**Comment by larkost** [comment]
But since we do have the `final` function, people will expect that if they have that for the callback style, that it would resolve the error. But your proposal to have the error also be expressed by a Promise rejection would necessarily mean that someone not using Promises would get an error in their code, even though they already handed it in their `final` function. Therefore we need to have the error go either to the `final` function (when there is one), or out as a Promise rejection (when there is not one), not both.

And again, I will point out that we are supporting returning Promises from the `row` handler. Not supporting that in the `final` handler just to keep it pure seems like the wrong solution. As I said: we have already committed to the mixed use case in this function. We need to see that through in as logical/least-surprise way as we can.

**Comment by danielmewes** [comment]
I had a brief look at this and these are my thoughts:

1. I agree with @marshall007 on the `row error passed through outer promise rather than final function` argument. I think errors should not only be passed to the `final` callback, but should also be rejected on the returned promise. As far as I understand there isn't any issue with an unhandled rejected promise being returned if it is never used. There shouldn't be a need to any catch blocks if you don't intent to use the returned promise, unless I'm completely wrong on how this works.

2. I agree with @larkost that we should wait on a promise returned by the `final` function. I think the distinction between callback-style and promise-style isn't as clear as it might at first seem. `eachAsync` is always a partially callback-style command, since it calls a callback for every iterated row.
I can see the argument of taking a variant of `eachAsync` that's designed around promises, and then nodeifying it in order to get a purely callback-based alternative from it. However I don't think that there's anything wrong if a user wants to use the `final` callback and also wants to use the returned promise.
I don't think it would be obvious that the `final` callback is merely intended for use in a purely callback-style environment. Sure, every case where `final` is used with promises can be rewritten to a `then` block on the promise returned from `eachAsync`. But again I don't think this will necessarily be obvious, and some users might end up using the `final` callback anyway in promise-based code. Additionally there's the argument from symmetry with the row handler callback. The motivation for the row handler callback and the `final` callback existing in the first place are different, but that doesn't mean that they should behave differently.

Hope that makes sense.

**Comment by larkost** [comment]
The rejection is not silently ignored (thankfully for my sanity in debugging):
```
> var Promise = require("bluebird");
undefined
> Promise.delay(10).then(function() { return Promise.reject() })
{ _bitField: 0,
  _fulfillmentHandler0: undefined,
  _rejectionHandler0: undefined,
  _promise0: undefined,
  _receiver0: undefined,
  _settledValue: undefined,
  _boundTo: undefined }
> Possibly unhandled Error: undefined
    at Function.Promise$Reject (*snip*/bluebird/js/main/promise.js:300:21)
    at repl:1:53
    at tryCatch1 (*snip*/bluebird/js/main/util.js:43:21)
    at Promise$_callHandler [as _callHandler] (*snip*/bluebird/js/main/promise.js:649:13)
    at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (*snip*/bluebird/js/main/promise.js:663:18)
    at Promise$_settlePromiseAt [as _settlePromiseAt] (*snip*/bluebird/js/main/promise.js:826:14)
    at Promise$_settlePromises [as _settlePromises] (*snip*/bluebird/js/main/promise.js:969:14)
    at Async$_consumeFunctionBuffer [as _consumeFunctionBuffer] (*snip*/bluebird/js/main/async.js:75:12)
    at Async$consumeFunctionBuffer (*snip*/bluebird/js/main/async.js:38:14)
    at process._tickCallback (node.js:415:13)
```

**Comment by danielmewes** [comment]
@larkost I was just about to comment. @nighelles also pointed me towards this fact.

That changes my conclusion on 1, and it looks like there's no way we can still reject the promise when the `final` callback is used. @marshall007 is this correct?

**Comment by larkost** [comment]
The only exception to that would be if we have an error within the `final` handler. Then we have no other option (that the user could theoretically catch) than to put it out out the Promise chain as an error/rejection.

**Comment by marshall007** [comment]
It looks like using [`.nodeify(cb)`](http://bluebirdjs.com/docs/api/ascallback.html) (as we typically do to support callbacks) does not modify the promise yet also prevents "uncaught" rejections from bubbling up.

```
$ node
> var Promise = require('bluebird')
undefined
> Promise.reject('error').asCallback((err, result) => console.log(err))
Promise {
  _bitField: 16777216,
  _fulfillmentHandler0: 'error',
  _rejectionHandler0: undefined,
  _promise0: undefined,
  _receiver0: undefined }
error
> Promise.reject('error').nodeify((err, result) => console.log(err)).catch((err) => console.log(err))
Promise {
  _bitField: 0,
  _fulfillmentHandler0: undefined,
  _rejectionHandler0: undefined,
  _promise0: undefined,
  _receiver0: undefined }
error
error
```

I still think resolving promises from `final`doesn't make any sense. For one thing, there's no reason for promise people to use the `final` callback. It can already be accomplished by doing `eachAsync(...).then(() => Promise.delay(10))`, and this is how you would expect a Promise-based API to behave. Another is that it it results in discrepancies between the callback and Promise based APIs. There currently is no callback equivalent to returning a promise from `final`.

If you look at the code, our current implementation is:

```js
return nextCb().nodeify(errCb)
```

What's being proposed (resolving promises from `final` and thus having side-effects on the outer promise) would look something like:

```js
return nextCb() if not errCb?
return nextCb().then ->
  return Promise.resolve(errCb())
.catch (err) ->
  return Promise.resolve(errCb(err))
```

I think this helps illustrate why I think that what's being proposed is unconventional.

**Comment by danielmewes** [comment]
Thanks for the explanations @marshall007 .

This project is getting a bit out of hand.
To be honest I think in reality nobody is going to care about these edge cases of `eachAsync`, and we have enough other important things to take care of.

As a third option: How about we simply return `undefined` from `eachAsync` if the `final` callback is specified?

It would enforce the usage of *either* the promise-style or callback style final/error handler. It would completely avoid any ambiguity about whether promises returned from the `final` callback should be handled, or whether an error passed into the `final` callback should also be passed to the returned promise.

In that case, I also think that we should entirely ignore the return value of the `final` callback.

Would there be anything wrong with that API?

**Comment by danielmewes** [comment]
> This project is getting a bit out of hand.
> To be honest I think in reality nobody is going to care about these edge cases of eachAsync, and we have enough other important things to take care of.

Just to clarify: this didn't go against you @marshall007 , and I really appreciate your input on this. It was a general observation based on the fact that we have spent considerable time on discussing edge cases of eachAsync like these offline in the office. In retrospect we probably should just have sticked to your initial implementation that relied on the Bluebird defaults and defined the semantics of eachAsync accordingly...

**Comment by marshall007** [comment]
@danielmewes I'm happy with the third option you suggested, I don't see any problems with that. Apologies for all the back-and-forth. Hopefully we can get this wrapped up now.

**Comment by danielmewes** [comment]
Cool, let's go with option 3 then (returning `undefined` from `eachAsync` if `final` is provided). :-)
Pinging @nighelles.

**Comment by nighelles** [comment]
Ok, sounds good.

**Comment by larkost** [comment]
Sine there is therefor no way to catch any errors that come up from the `final` handler, are we going to just silently drop any (non Promise) values returned from that handler? Or are we going to just throw from an uncatchable spot?

**Comment by danielmewes** [comment]
I think it's fine to drop them (*including* promises I assume?).

@marshall007 any opinions on this one?

**Comment by marshall007** [comment]
@larkost I think the plan is to ignore the return value of `final` entirely, Promise or not. Any exceptions thrown in `final` will automatically bubble up as uncaught exceptions, just as they would in any other callback.

**Comment by larkost** [comment]
As I understand it the plan is still to accept `Promise` returns, for consistency with the `row` handler. Even if we don't, we should still warn people if we don't accept them, as otherwise people will run into bad surprises when their code in production does not work like it happened to in development because we are not actually following returned `Promise`s.

**Comment by danielmewes** [comment]
@larkost That was a misunderstanding then. If we don't return anything from `eachAsync` when it has a `final` callback, there is nothing we could do with those promises.
People can't run into bad surprises as far as I can tell, because it makes no difference whether we resolve the promise or not. We will return `undefined` in either case.

**Comment by larkost** [comment]
I am going to close this, since my work on this is done, and has been folded into @nighelles work on the driver in issue #5518.
